foreshadow/cleaners/data_cleaner.py

```diff
@@ -1,26 +1,279 @@
 """Cleaner module for cleaning data as step in Foreshadow workflow."""
+from collections import namedtuple
+
+import pandas as pd
 from sklearn.base import BaseEstimator, TransformerMixin
 
+from foreshadow.core.base import PreparerStep
+from foreshadow.exceptions import InvalidDataFrame, SmartResolveError
+from foreshadow.metrics.internals import avg_col_regex, regex_rows
+from foreshadow.transformers.core import SmartTransformer
+from foreshadow.transformers.core.wrapper import _Empty
+from foreshadow.utils.testing import dynamic_import
+from foreshadow.utils.validation import check_df
+from foreshadow.transformers.core.notransform import NoTransform
+
+
+CleanerReturn = namedtuple("CleanerReturn", ["row", "match_lens"])
+
+
+class DataCleaner(PreparerStep):
+    """Determine and perform best data cleaning step."""
+
+    def __init__(self, *args, **kwargs):
+        """Define the single step for DataCleaner, using SmartCleaner.
+
+        Args:
+            *args: args to PreparerStep constructor.
+            **kwargs: kwargs to PreparerStep constructor.
+
+        """
+        super().__init__(*args, use_single_pipeline=True, **kwargs)
+
+    def get_mapping(self, X):
+        return self.separate_cols(
+            transformers=[[SmartFlatten(), SmartCleaner()] for c in X],
+            X=X
+        )
+
+
+class SmartCleaner(SmartTransformer):
+    """Intelligently decide which cleaning function should be applied."""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def pick_transformer(self, X, y=None, **fit_params):
+        """Get best transformer for a given column.
+
+        Args:
+            X: input DataFrame
+            y: input labels
+            **fit_params: fit_params
+
+        Returns:
+            Best data cleaning transformer.
+
+        """
+        from foreshadow.cleaners.internals import __all__ as cleaners
+
+        cleaners = [
+            (dynamic_import(cleaner, "foreshadow.cleaners.internals"),
+             cleaner)
+            for cleaner in cleaners
+            if cleaner.lower().find("cleaner") != -1
+        ]
+        best_score = 0
+        best_cleaner = None
+        for cleaner, name in cleaners:
+            cleaner = cleaner(name=name)
+            score = cleaner.metric_score(X)
+            if score > best_score:
+                best_score = score
+                best_cleaner = cleaner
+
+        if best_cleaner is None:
+            return NoTransform()
+        return best_cleaner
 
-class DataCleaner(BaseEstimator, TransformerMixin):
-    """Wrapper class to determine and perform best data cleaning step."""
 
+class SmartFlatten(SmartTransformer):
```

Comment:
I might add a class doc regardless to talk about the general use for SmartFlatten